sqlstats: fix sampling logic for first recording of a statement

xinhaoz opened the pull request:
By default we sample execution stats for 1% of statements,
(`sql.txn_stats.sample_rate`). We also turn on sampling if
this is the first time we're recording this fingerprint in the
application container. This 'sample on first encounter' logic did
not factor in the scenario where we may be unable to write new
fingerprints due to reaching max container capacities. Previously,
this meant that once the containers were full we'd begin to trace
every fingerprint that did not already get recorded.

This commit ensures that we also check the capacity when deciding
to sample a fingerprint we haven't recorded yet, skipping the sampling
if it is full.  Note that although the container may reach capacity
during the statement's execution, it's more important that we stop sampling
once we are sure the fingerprint cannot be written to the container.

Fixes: https://github.com/cockroachdb/cockroach/issues/123690

Release note: None

xinhaoz committed:
sqlstats: fix sampling logic for first recording of a statement

By default we sample execution stats for 1% of statements,
(`sql.txn_stats.sample_rate`). We also turn on sampling if
this is the first time we're recording this fingerprint in the
application container. This 'sample on first encounter' logic did
not factor in the scenario where we may be unable to write new
fingerprints due to reaching max container capacities. Previously,
this meant that once the containers were full we'd begin to trace
every fingerprint that did not already get recorded.

This commit ensures that we also check the capacity when deciding
to sample a fingerprint we haven't recorded yet, skipping the sampling
if it is full.  Note that although the container may reach capacity
during the statement's execution, it's more important that we stop sampling
once we are sure the fingerprint cannot be written to the container.

Fixes: #123690

Release note: None

xinhaoz commented:
need to add testing

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/123696)
<!-- Reviewable:end -->

xinhaoz committed:
sql: add sqlstats.RecordedTxnStats to OnRecordTxnFinish test hook

Add the recorded txn stats so we can do some testing validation
on the stats collected.

Epic: none

Release note: None

xinhaoz committed:
sql: simplify the assignment of `shouldCollectTxnExecutionStats`

This commit simplifies the assignment of this field on the conn
executor's `extraTxnState`.

Epic: none

Release note: None

dhartunian commented on pkg/sql/instrumentation.go:
```diff
@@ -470,20 +470,20 @@ func (ih *instrumentationHelper) Setup(
-
-	// Don't collect it if Stats Collection is disabled. If it is disabled the
-	// stats are not stored, so it always returns false for previouslySampled.
-	// TODO(117690): Unify StmtStatsEnable and TxnStatsEnable into a single cluster setting.
```
Should this TODO get deleted?

xinhaoz replied:
Oops, I'll put this back. It still looks relevant.

dhartunian commented on pkg/sql/instrumentation.go:
I'm a little confused about how `limitReached` is supposed to affect `ih.collectExecStats`. If we're at the limit shouldn't we override `collectTxnExecStats` if it's set to `true`? Or is the assumption that it's set to true only if it's a collected fingerprint already?

xinhaoz replied:
There's no assumption of that sort here. We're using `limitReached` specifically just to correct the case where we want to override `collectTxnExecStats` because the container hasn't recorded this fingerprint yet since if it's full it'll never insert this new fingerprint. `collectTxnExecStats` could be set for a statement that's in or not in the container.

If we'd prefer to only set `collectTxnExecStats` for statements in the container, we can investigate the implications of that. The only reason I didn't pursue that is I'm not sure if this sampling logic could be used outside of sql stats for any other reporting / metrics.

abarganier reviewed (approved):
Looks like this will fix the bug, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

I have a small readability nit, feel free to take it or leave it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123696#-:-NxYEix23VxQL2q1CcV4:b-sy44t8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @xinhaoz)

___
*[`pkg/sql/instrumentation.go` line 473 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/123696#-NxYAg_dDIIF8EDgfA-K:-NxYAg_dDIIF8EDgfA-L:b71m5me) ([raw file](https://github.com/cockroachdb/cockroach/blob/6abc367f91f82eb3286f624274278ef1d526f0c7/pkg/sql/instrumentation.go#L473)):*
> ```Go
> 	}
> 
> 	shouldSampleFirstEncounter := func() bool {
> ```

nit: The checks here LGTM, but can we make this more readable with some early returns? 

For example, it seems like if `statsCollectionEnabled == false`, or `limitReached == false`, we'll never indicate that we should sample. So, we can return early in those cases. It would help simplify the conditional. `&&`ing many things together is the same in practice, but can be more difficult for readers to wrap their head around. 

As a small bonus, it speeds things up a tiny bit by avoiding unnecessary reads of cluster settings/atomic ints/etc when we already know the answer based on a single variable.

Factoring this into a function was a great idea for readability, nice work.



_Code quote:_
```Go
	shouldSampleFirstEncounter := func() bool {
```
___
*[`pkg/sql/instrumentation.go` line 480 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/123696#gh-1596029237:-NxYDeUh62OiPDPKeU2I:b-26qrjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6abc367f91f82eb3286f624274278ef1d526f0c7/pkg/sql/instrumentation.go#L480)):*
> ih.collectExecStats = collectTxnExecStats || shouldSampleFirstEncounter()

Giving precedence to `collectTxnExecStats` makes sense to me. We defer to probabilistic sampling. If our sample rate didn't choose this txn for sampling, then we check `shouldSampleFirstEncounter()` to see if we should sample anyway, based on the criteria defined within that function (which protects against the bug being solved here). 


<!-- Sent from Reviewable.io -->

xinhaoz commented:
TFTR! Cleaned up and split the conditional.
bors r+
